adam_core/utils/indexable.py

```diff
@@ -1,174 +1,451 @@
+    from astropy.time import Time
+    from astropy import units as u
+
+    # Instantiate an Orbits object with 3 orbits
+    t0 = Time([59000.0], scale="utc", format="mjd")
+    orbits = Orbits.from_horizons(["Eros", "Ceres", "Duende"], t0)
+    len(orbits) == 3
+
+    # If I only wanted one orbit
+    single_orbit = orbits[0] # Orbits with one orbit
+
+    # Lets remove the other ones
+    del orbits[1:]
+
+    # Lets propagate the orbits to 10 new times
+    orbits = Orbits.from_horizons(["Eros", "Ceres", "Duende"], t0)
+    t1 = t0 + np.arange(0, 10) * u.day
+    backend = PYOORB()
+    propagated_orbits = backend.propagate_orbits(orbits, t1)
+
+    # The underlying arrays are now 10 * 3 in length
+    len(propagated_orbits) == 30
+
+    # If we update the index to be on orbit_id
+    # then the length of the class is 3 even though each
+    # member is 10 * 3 in length
+    propagated_orbits.set_index("orbit_id")
+    len(propagated_orbits) == 3
+
+    # Lets delete all of the states for "Duende"
+    del propagated_orbits[2]
+    ```
+    """
+
+    def __init__(self, index_values: Optional[Union[str, npt.ArrayLike]] = None):
+        self.set_index(index_values)
+        return
 
     @property
     def index(self):
-        return self._index
+        """
+        The externally facing index of the class.
+        """
+        return self._class_index
 
-    def set_index(self, index: Union[str, np.ndarray]):
-        if isinstance(index, str):
-            class_index = getattr(self, index)
-            self._index_attribute = index
-            array_index = np.arange(0, len(class_index), dtype=int)
+    @index.setter
+    def index(self, value):
+        self.set_index(value)
 
-        elif isinstance(index, np.ndarray):
-            class_index = index
-            self._index_attribute = None
-            array_index = np.arange(0, len(index), dtype=int)
+    @index.deleter
+    def index(self):
+        self.set_index()
+
+    def _check_member_validity(self):
+        """
+        Scans the class's sliceable members and raises an error if they are
+        not all of the same length.
+
+        Sets
+        ----
+        self._member_length : int
+            The length of the members of the class.
+        self._member_index : `~numpy.ndarray`
+            The index of the members of the class.
+
+        Raises
+        ------
+        ValueError: If the members are not all of the same length.
+        """
+        # Scan members and if they are of a type that is sliceable
+        # then store the length. All sliceable members should have the same length
+        # along their first axis.
+        member_lengths = []
+        for k, v in self.__dict__.items():
+            if isinstance(v, SLICEABLE_DATA_STRUCTURES):
+                member_lengths.append(len(v))
 
+        if len(np.unique(member_lengths)) != 1:
+            raise ValueError("All sliceable members must have the same length.")
```

Comment:
This looks like a job for the Python builtin container `set`! It has O(1) insertion and uniqueness checks. Of course, the number of members will always be tiny, but it's worth knowing about as a faster and more idiomatic way to keep track of unique values.

```suggestion
        member_lengths = {len(v) for v in self.__dict__.values() if isinstance(v, SLICEABLE_DATA_STRUCTURES)}

        if len(member_lengths) != 1:
            raise ValueError("All sliceable members must have the same length.")
        member_length = member_lengths.pop()
```

Reply:
Done!